File: prescriptions/rsa_generator.go
package prescriptions

import (
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"errors"
	"fmt"
	"time"

	"doctor-service/entities"

	"github.com/google/uuid"
)

type RSADigitalPrescription struct {
	PrescriptionID   uuid.UUID `json:"prescription_id"`
	DoctorID         uuid.UUID `json:"doctor_id"`
	PatientID        string    `json:"patient_id"`
	Diagnosis        string    `json:"diagnosis"`
	Medicines        []string  `json:"medicines"`
	DigitalSignature string    `json:"digital_signature"` // Base64 encoded RSA-SHA256 signature
	PublicKeyPEM     string    `json:"public_key_pem"`     // Public Key for external verification
	QRPayload        string    `json:"qr_payload"`
	IssuedAt         time.Time `json:"issued_at"`
}

type RSAPrescriptionGenerator struct {
	privateKey *rsa.PrivateKey
	publicKey  *rsa.PublicKey
}

func NewRSAPrescriptionGenerator() (*RSAPrescriptionGenerator, error) {
	key, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		return nil, fmt.Errorf("failed to generate RSA keypair: %w", err)
	}
	return &RSAPrescriptionGenerator{
		privateKey: key,
		publicKey:  &key.PublicKey,
	}, nil
}

func (g *RSAPrescriptionGenerator) IssuePrescription(doctor *entities.Doctor, patientID, diagnosis string, medicines []string) (*RSADigitalPrescription, error) {
	if doctor.Status != entities.DoctorStatusVerified || !doctor.PrescriptionEnabled {
		return nil, fmt.Errorf("doctor '%s' is not authorized to issue digital prescriptions", doctor.PublicID)
	}

	prescriptionID := uuid.New()
	issuedAt := time.Now()
	medsJSON, _ := json.Marshal(medicines)

	// 1. Signature Input
	signatureInput := fmt.Sprintf("%s|%s|%s|%s|%s|%s",
		prescriptionID.String(),
		doctor.PublicID.String(),
		patientID,
		diagnosis,
		string(medsJSON),
		issuedAt.Format(time.RFC3339),
	)

	// 2. Sign using RSA-SHA256 Private Key
	hashed := sha256.Sum256([]byte(signatureInput))
	signatureBytes, err := rsa.SignPKCS1v15(rand.Reader, g.privateKey, crypto.SHA256, hashed[:])
	if err != nil {
		return nil, fmt.Errorf("RSA signature generation failed: %w", err)
	}
	digitalSignature := base64.StdEncoding.EncodeToString(signatureBytes)

	// Export Public Key as PEM
	pubBytes, err := x509.MarshalPKIXPublicKey(g.publicKey)
	var pubPEM string
	if err == nil {
		pubBlock := &pem.Block{Type: "PUBLIC KEY", Bytes: pubBytes}
		pubPEM = string(pem.EncodeToMemory(pubBlock))
	}

	// 3. Generate QR Payload
	qrData := map[string]interface{}{
		"pid":  prescriptionID.String(),
		"doc":  doctor.PublicID.String(),
		"name": doctor.FirstName + " " + doctor.LastName,
		"pat":  patientID,
		"sig":  digitalSignature[:24],
		"ts":   issuedAt.Unix(),
		"url":  fmt.Sprintf("https://practo-doctor.portal/prescriptions/%s/verify", prescriptionID.String()),
	}

	qrJSON, _ := json.Marshal(qrData)
	qrPayload := base64.StdEncoding.EncodeToString(qrJSON)

	return &RSADigitalPrescription{
		PrescriptionID:   prescriptionID,
		DoctorID:         doctor.PublicID,
		PatientID:        patientID,
		Diagnosis:        diagnosis,
		Medicines:        medicines,
		DigitalSignature: digitalSignature,
		PublicKeyPEM:     pubPEM,
		QRPayload:        qrPayload,
		IssuedAt:         issuedAt,
	}, nil
}

func VerifyRSASignature(p *RSADigitalPrescription) (bool, error) {
	if p.PublicKeyPEM == "" || p.DigitalSignature == "" {
		return false, errors.New("missing public key or digital signature")
	}

	block, _ := pem.Decode([]byte(p.PublicKeyPEM))
	if block == nil {
		return false, errors.New("failed to decode public key PEM")
	}

	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return false, fmt.Errorf("failed to parse public key: %w", err)
	}
	rsaPub, ok := pub.(*rsa.PublicKey)
	if !ok {
		return false, errors.New("invalid RSA public key type")
	}

	medsJSON, _ := json.Marshal(p.Medicines)
	signatureInput := fmt.Sprintf("%s|%s|%s|%s|%s|%s",
		p.PrescriptionID.String(),
		p.DoctorID.String(),
		p.PatientID,
		p.Diagnosis,
		string(medsJSON),
		p.IssuedAt.Format(time.RFC3339),
	)

	hashed := sha256.Sum256([]byte(signatureInput))
	sigBytes, err := base64.StdEncoding.DecodeString(p.DigitalSignature)
	if err != nil {
		return false, fmt.Errorf("failed to decode signature base64: %w", err)
	}

	err = rsa.VerifyPKCS1v15(rsaPub, crypto.SHA256, hashed[:], sigBytes)
	if err != nil {
		return false, nil // Invalid signature
	}

	return true, nil
}

File: python_backend/prescriptions/rsa_generator.py
import json
import base64
import uuid
from datetime import datetime
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives import serialization

class RSADigitalPrescription:
    def __init__(self, prescription_id, doctor_id, patient_id, diagnosis, medicines, digital_signature, public_key_pem, qr_payload, issued_at):
        self.prescription_id = prescription_id
        self.doctor_id = doctor_id
        self.patient_id = patient_id
        self.diagnosis = diagnosis
        self.medicines = medicines
        self.digital_signature = digital_signature
        self.public_key_pem = public_key_pem
        self.qr_payload = qr_payload
        self.issued_at = issued_at

class RSAPrescriptionGenerator:
    def __init__(self):
        self.private_key = rsa.generate_private_key(
            public_exponent=65537,
            key_size=2048,
        )
        self.public_key = self.private_key.public_key()

    def issue_prescription(self, doctor, patient_id: str, diagnosis: str, medicines: list) -> RSADigitalPrescription:
        if doctor.status != "VERIFIED" or not doctor.prescription_enabled:
            raise ValueError(f"doctor '{doctor.public_id}' is not authorized to issue digital prescriptions")

        prescription_id = uuid.uuid4()
        issued_at = datetime.now()
        meds_json = json.dumps(medicines, separators=(',', ':'))

        signature_input = f"{prescription_id}|{doctor.public_id}|{patient_id}|{diagnosis}|{meds_json}|{issued_at.isoformat()}Z"

        signature_bytes = self.private_key.sign(
            signature_input.encode('utf-8'),
            padding.PKCS1v15(),
            hashes.SHA256()
        )
        digital_signature = base64.b64encode(signature_bytes).decode('utf-8')

        pub_pem = self.public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo
        ).decode('utf-8')

        qr_data = {
            "pid": str(prescription_id),
            "doc": str(doctor.public_id),
            "name": f"{doctor.first_name} {doctor.last_name}",
            "pat": patient_id,
            "sig": digital_signature[:24],
            "ts": int(issued_at.timestamp()),
            "url": f"https://practo-doctor.portal/prescriptions/{prescription_id}/verify"
        }

        qr_json = json.dumps(qr_data, separators=(',', ':'))
        qr_payload = base64.b64encode(qr_json.encode('utf-8')).decode('utf-8')

        return RSADigitalPrescription(
            prescription_id=prescription_id,
            doctor_id=doctor.public_id,
            patient_id=patient_id,
            diagnosis=diagnosis,
            medicines=medicines,
            digital_signature=digital_signature,
            public_key_pem=pub_pem,
            qr_payload=qr_payload,
            issued_at=issued_at
        )

def verify_rsa_signature(p: RSADigitalPrescription) -> bool:
    if not p.public_key_pem or not p.digital_signature:
        raise ValueError("missing public key or digital signature")

    try:
        public_key = serialization.load_pem_public_key(
            p.public_key_pem.encode('utf-8')
        )
    except Exception as e:
        raise ValueError(f"failed to parse public key: {e}")

    meds_json = json.dumps(p.medicines, separators=(',', ':'))
    signature_input = f"{p.prescription_id}|{p.doctor_id}|{p.patient_id}|{p.diagnosis}|{meds_json}|{p.issued_at.isoformat()}Z"

    try:
        sig_bytes = base64.b64decode(p.digital_signature)
    except Exception as e:
        raise ValueError(f"failed to decode signature base64: {e}")

    try:
        public_key.verify(
            sig_bytes,
            signature_input.encode('utf-8'),
            padding.PKCS1v15(),
            hashes.SHA256()
        )
        return True
    except Exception:
        return False
